### PR TITLE
Delete unnecessary checks before some function calls

### PR DESCRIPTION
**Pull request by elfring**
* [The function "free"](http://stackoverflow.com/questions/18775608/free-a-null-pointer-anyway-or-check-first "Free a null pointer anyway or check first?") is documented in the way that no action shall occur for a passed null pointer. Redundant safety checks can be avoided.
* The following functions perform also input parameter validation.
  * [Curl_cookie_cleanup](https://github.com/bagder/curl/blob/df5578a7a304a23f9aa3670daff8573ec3ef416f/lib/cookie.c#L1207 "Cookie clean-up")
  * [curl_formfree](https://github.com/bagder/curl/blob/df5578a7a304a23f9aa3670daff8573ec3ef416f/lib/formdata.c#L956 "Release a form post chain.")

  It is therefore not needed that a function caller repeats a corresponding check.

### CODE REVIEW

**Comment by captain-caveman2k** [comment]
I appreciate what the spec says about not needing to check the value of the pointer but, Nick has already mentioned that some systems may crash on free(NULL) so do we really want to remove these checks?

In my opinion it may be worth:

* Removing the checks as suggested
* Replacing free() with Curl_safefree()
* Modifying Curl_safefree() to have the check in (as it currently does) on platforms with known issues and without the check on other platforms.

The other benefits of using Curl_safefree() are a) the integration with the curl memory tracking routines as well as b) automatically setting the value of the pointer to NULL afterwards - thus removing the need to perform "pointer = NULL;" in various places ;-)

**Comment by bagder** [comment]
I think I would be fine with saying that we require a free() function that can deal with NULL in itself. If anyone would build curl on a system where that isn't true, it wouldn't be that terribly hard to provide a wrapper function that does the necessary checks.

**Comment by captain-caveman2k** [comment]
Maybe I've misunderstand something but we already have that function (well, it's a #define) - Curl_safefree() ;-)

**Comment by bagder** [comment]
Yeah we do, I'm just not convinced I want every free() in the code converted to a Curl_safefree() and if we don't do that replace-all, we will end up with a mix of them... and then, going forward, I think we will end up with a place in the code where free() will be a called with a NULL pointer at some point in time.

**Comment by captain-caveman2k** [comment]
But we have that mix already :(

From a quick scan of the source code, we have 605 uses of Curl_safefree() and 378 of free().

**Comment by bagder** [comment]
We have the mix, but we currently avoid free(NULL) widely so the mix is not a problem.

Hence my position that I think we should just require a free() that accepts NULL because then we can clean up the code as per this pull request _and_ the mix of free/safefree is not a problem.

**Comment by captain-caveman2k** [comment]
Now I'm well and truly confused :(

You seem to say that currently avoiding free(NULL), the mix is not a problem but then go onto say that with support for this pull request and by using free() that accepts NULL the mix isn't a problem.

Are you suggesting me introduce our own version of free() that accepts NULL?

Sorry if I'm missing something obvious - I'm not normally this stupid as you know ;-)

**Comment by bagder** [comment]
Ok, I'm probably clear as mud. Let me try again. Here's what I say:

TODAY: we can (probably) build and run on systems with a free() function that can not handle NULL because we avoid calling free() like that. The current mix of free and safefree is not a problem, as none makes any free(NULL) calls because we have checks to avoid that.

FUTURE: if we want to cleanup the code as per this pull-request and avoid if()s before the free() calls (which I think makes sense), and *IF* we want to continue to support that "broken" sort of free() implementation that doesn't handle NULL nicely, we pretty much HAVE TO replace all uses of free() with safefree(). An operation I rather not do.

Therefore. If we accept this pull-request, I think we should also say that we *require* free() to handle NULL because we don't want to do search-replace of all free() calls. I don't think this is a high price to pay.

**Comment by jay** [comment]
+1 for allowing free(NULL). If my grasp of this issue is correct I believe it could be done easily. Here's how I would do it:
https://github.com/bagder/curl/compare/master...jay:rpl_free

**Comment by elfring** [comment]
How do you think about cooperation with configuration checks like [the macro call "`AC_FUNC_MALLOC`"](http://www.gnu.org/software/autoconf/manual/autoconf-2.67/html_node/Particular-Functions.html "Checking memory management interface?") in your build specification?
Would you like to [reuse function wrappers](http://stackoverflow.com/questions/262439/create-a-wrapper-function-for-malloc-and-free-in-c "Create a wrapper function for malloc and free in C?") under specific circumstances?

**Comment by bagder** [comment]
@elfring how would that help for this case?

**Comment by elfring** [comment]
Do you want to detect standard-incompliant function implementations for the discussed memory management by corresponding configuration checks automatically?

**Comment by bagder** [comment]
@elfring I find it unlikely that the systems with non-compliant free() implementations will run configure in the first place.

I also think we can just document our requirement that free() should work with NULL input in the INTERNALS document and be done with it. Then if someone wants to run libcurl on a system with a non-compliant free() the work is left for that person/team and the fix for that could then very well be in the style @jay showed above. Let's not overwork this.

**Comment by captain-caveman2k** [comment]
@bagder - Ah yes, that explanation makes sense now ;-)

Do we know what the buggy platforms, for sure, are?

ie. As soon as we release 7.42.0, with this PR in, are we going to create a world of pain for someone? I honestly don't know the answer to that but if there is a slimmest of possibilities that we may do I think we should either do the work, similar to what Jay mentioned, or use Curl_safefree() instead.

I'm going off on a tangent here but if we don't use Curl_safefree() then perhaps we should consider deprecating it and using free() all over the place. I personally would like to see one or the other used and don't really have any preference except our version saves assigning null afterwards ;-) 

Anyway back to topic... after months of breakages on platforms that don't have 64-bit support, I don't really want to break things again one release later - even for a two or three platforms :(

Another thing I wanted to ask... are there any performance benefits or disadvantages or using if(ptr) free(ptr) - ie, ptr doesn't need to be pushed onto the stack vs. the overhead of the if condition?

**Comment by bagder** [comment]
I don't know off the top of my head which platforms they are, but I'm quite sure none of those are among our top-20 platforms. And even if someone does get bitten by this, it will not leave them in a world of pain as the work-around they can do is quite simple and mostly already done by @jay above.

I can't think of much of a performance diff with if(x) free() in our code compared to that same check within the free() implementation instead. On the contrary, I think most of those if()s are done on valid pointers so by removing if()s we will actually run less code than today - on average. Hardly measurable...

About the safefree / free mix, we can deal with that separately later on but I would prefer to go toward plain old free() (since that is what people know and are used to) rather than the opposite, and when we allow NULL sent to free(), we can also skip that check in the safefree macro!

**Comment by elfring** [comment]
Would effects like the following be useful here?
* Reduction of [object code size](https://en.wikipedia.org/wiki/Object_code "Description for object code")
* Less source code for review
* Less conditional jumps which influence [the branch prediction by processors](https://en.wikipedia.org/wiki/Branch_predictor "Description for branch prediction")
* [Inlining opportunities](https://en.wikipedia.org/wiki/Inline_expansion "Description for inline expansion")
* [Cache locality](https://en.wikipedia.org/wiki/Locality_of_reference "Description for locality of reference")

Would anybody like to measure concrete differences?

**Comment by bagder** [comment]
@elfring thanks but I don't think that's necessary. What we debate here is rather what functionality we should maintain and what not and how to proceed. I don't think anyone here doubts the usefulness of removing unnecessary checks, the question is what price we pay and if it is worth it. I wanted to give people time to respond before I proceed but I don't see anyone objecting strongly which will make me merge this.

**Comment by bagder** [comment]
Thanks a lot, both patches have been merged as of commit 9e661601feba.
